### ISSUE TITLE
Relax the restrictions on 64-bit PETSc.

### ISSUE THREAD

**Issue by drwells** [dealii/dealii#16528]
This is from slack.

Since `PetscInt` is a signed integer, we should always
1. make a copy (no `reinterpret_cast<PetscInt *>`)
2. do a range check (since, for 32-bit indices, the maximum range is significantly different)

If we're doing both, then we don't need to restrict ourselves to matching `DEAL_II_WITH_64BIT_INDICES` and `sizeof(PetscInt) == 8`. That would simplify the CI set up a bit (and make things easier for users).

**Comment by bangerth** [comment]
Perhaps via a `convert_to_PetscInt()` function, rather than always doing it by hand in all places necessary?

**Comment by drwells** [comment]
That's a good idea

**Comment by jpthiele** [comment]
We discussed this yesterday regarding Trilinos/Tpetra. 
I noticed #17481 was missing from this issue, any other steps/PRs?

So to summarize (and please correct me if I got anything wrong)
- The new type PetscInt stores the type of the PETSc installation
- This type is used anywhere when dealing with PETSc Linear Algebra classes
- deal.II ordinates, e.g. from DoFHandler, will be converted/cast to PetscInt inside the PETSc wrappers
- A new assertion checks if this conversion was valid

What I am wondering is whether the conversion is happy when a 32Bit deal.II global index is converted to PETSc with 64 Bit?
And vice versa, will a 64Bit stored deal.II global index be an accepted conversion to 32Bit PetscInt if it lies in the 32Bit interval?

**Comment by bangerth** [comment]
`PetscInt` is a type that PETSc itself exports (and has always exported). It's not something we declare ourselves. We need to use it in all places where we interact with PETSc, for example where we need to pass an index to PETSc.

**Comment by jpthiele** [comment]
Ah okay, since Tpetra is theoretically templated (but allows only one instantiation) we define `GO` in [`trilinos_tpetra_types`](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/trilinos_tpetra_types.h).
This matches `dealii::types::signed_global_dof_index`, which was introduced for Tpetra and is currently defined based on `DEAL_II_WITH_64_BIT_INDICES` but could be removed again in favor of `GO` which we could rename to something more clear, e.g. `TpetraTypes::global_dof_index` or `signed_global_dof_index` and then based on `TPETRA_HAVE_INST_INT_LONG_LONG` or `TPETRA_HAVE_INST_INT_INT`.

**Comment by bangerth** [comment]
Yes. Conceptually, deal.II should be able to choose its own `types::global_dof_index` (and, if we need that somewhere, the corresponding `types::signed_global_dof_index`), and that should be independent of `PetscInt` and `TpetraTypes::global_dof_index = GO` as long as one can map one to the other; in practice, we want `PetscInt` and `GO` to be at least as big as `types::global_dof_index` (perhaps allowing for a difference in signedness -- the factor of 2 we lose in the size is probably not important).

**Comment by jpthiele** [comment]
The second the factor of 2, I don't think I have so far heard of anyone needing a quadrillion DoFs and signed 64 Bit is larger than 9 quintillion.

**Comment by tamiko** [comment]
Let's tackle this early after the release.
